### PR TITLE
Add Filter&Alarm !Exists For AWS MC Auth Failures

### LINKED ISSUE #1199: Add Filter And Alarm Not Exists For AWS Management Console Authentication Failures query for CloudFormation
### Platform
CloudFormation

### Provider
AWS

### Description
A log metric filter and alarm should exist for AWS Management Console Authentication Failures

### PR DESCRIPTION
**Pull request by fabioGoncalvesCx**
Added new query for AWS CloudFormation.

Provide a CloudFormation resource and a resource to ensure that AWS Management Console Authentication Failures are catched by a log metric filter and an alarm

Closes #1199

### CODE REVIEW

**Review by ruigomescx — APPROVED** [review]

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/metadata.json [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{
+  "id": "Filter_And_Alarm_Exist_For_AWS_Management_Console_Authentication_Failures",
+  "queryName": "Filter And Alarm Not Exists For AWS Management Console Authentication Failures",
+  "severity": "MEDIUM",
+  "category": "Monitoring",
+  "descriptionText": "A log metric filter and alarm should exist for AWS Management Console Authentication Failures",
+  "descriptionUrl": "https://kb.securestate.vmware.com/monitor-console-auth-failures.html"
```
Comment (ruigomescx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/metadata.json:7]: point to official aws cloudformation documentation

**Review by ruigomescx — DISMISSED** [review]

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/metadata.json (hunk 1 above)
Comment (fabioGoncalvesCx) [reply assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/metadata.json]: Done

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/metadata.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
+  "id": "Filter_And_Alarm_Exist_For_AWS_Management_Console_Authentication_Failures",
+  "queryName": "Filter And Alarm Not Exists For AWS Management Console Authentication Failures",
+  "severity": "MEDIUM",
+  "category": "Monitoring",
+  "descriptionText": "A log metric filter and alarm should exist for AWS Management Console Authentication Failures",
+  "descriptionUrl": "https://docs.aws.amazon.com/securityhub/latest/userguide/securityhub-cis-controls.html"
+}
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/metadata.json:8]: [As a good POSIX practice](https://unix.stackexchange.com/a/18789) add end-of-file newline

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/query.rego [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                "keyActualValue": 	sprintf("'Resources.%s.FilterPattern' does not have the right specification", [name])
+            }
+}
+checkAlarm(metricFilter, resourceAlarm) = true {
+    resourceAlarm.Namespace != metricFilter
+}
+
+checkTopic(resourceAlarm, nameSNS) = true {
+    resourceAlarm.Ref != nameSNS
+}
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/query.rego:74]: [As a good POSIX practice](https://unix.stackexchange.com/a/18789) add end-of-file newline

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/test/negative.yaml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  MetricFilter:
+    Type: 'AWS::Logs::MetricFilter'
+    Properties:
+      LogGroupName: ''
+      FilterPattern: >-
+        { ($.eventName = ConsoleLogin) && ($.errorMessage = "Failed authentication") }
+      MetricTransformations:
+        - MetricValue: '1'
+          MetricNamespace: CloudTrailMetrics
+          MetricName: IAMPolicyEventCount
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/test/negative.yaml:38]: [As a good POSIX practice](https://unix.stackexchange.com/a/18789) add end-of-file newline

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/test/positive.yaml [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+  MetricFilter3:
+    Type: 'AWS::Logs::MetricFilter'
+    Properties:
+      LogGroupName: ''
+      FilterPattern: >-
+        { ($.eventName = ConsoleLogin) && ($.errorMessage = "Failed authentication") }
+      MetricTransformations:
+        - MetricValue: '1'
+          MetricNamespace: CloudTrailMetrics
+          MetricName: IAMPolicyEventCount
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/test/positive.yaml:109]: [As a good POSIX practice](https://unix.stackexchange.com/a/18789) add end-of-file newline

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/query.rego [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                "issueType":		"IncorrectValue",
+                "keyExpectedValue":  sprintf("'Resources.%s.Properties.FilterPattern' should have the right specification')", [name]),
+                "keyActualValue": 	sprintf("'Resources.%s.FilterPattern' does not have the right specification", [name])
+            }
+}
+checkAlarm(metricFilter, resourceAlarm) = true {
+    resourceAlarm.Namespace != metricFilter
+}
+
+checkTopic(resourceAlarm, nameSNS) = true {
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/query.rego:72]: This function is only being used once and should not be declared, replace it for a direct comparison

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/query.rego [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+    result := {
+                "documentId": 		input.document[i].id,
+                "searchKey": 	     sprintf("Resources.%s.Properties.FilterPattern", [name]),
+                "issueType":		"IncorrectValue",
+                "keyExpectedValue":  sprintf("'Resources.%s.Properties.FilterPattern' should have the right specification')", [name]),
+                "keyActualValue": 	sprintf("'Resources.%s.FilterPattern' does not have the right specification", [name])
+            }
+}
+checkAlarm(metricFilter, resourceAlarm) = true {
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/query.rego:68]: This function is only being used once and should not be declared, replace it for a direct comparison

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/metadata.json [hunk 8]
```diff
@@ -0,0 +1,8 @@
+{
+  "id": "Filter_And_Alarm_Exist_For_AWS_Management_Console_Authentication_Failures",
+  "queryName": "Filter And Alarm Not Exists For AWS Management Console Authentication Failures",
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/metadata.json:3]: Change **directory name** and **id** to: `authentication_failure_without_alarms`
Change **query name** to: `Authentication Failure Without Alarms`

File: assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/test/positive.yaml [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        attaching/detaching policies from IAM users, roles or groups.
+      MetricName: IAMPolicyEventCount
+      Namespace: CloudTrailMetrics
+      Statistic: Sum
+      Period: 300
+      EvaluationPeriods: 1
+      Threshold: 1
+      ComparisonOperator: GreaterThanOrEqualToThreshold
+      AlarmActions:
+        - Ref: SnsTopicFake
```
Comment (rogeriopeixotocx) [inline assets/queries/cloudFormation/filter_and_alarm_exist_for_aws_management_console_authentication_failures/test/positive.yaml:98]: Either remove the reference (intrinsic function `Ref`) to Resource that doesn't exists or add it as a [Parameter](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html) to the header of the template

**Review by rogeriopeixotocx — CHANGES_REQUESTED** [review]
Requesting changes

**Comment by markmishaevcx** [comment]
Scan submitted to Checkmarx

**Comment by markmishaevcx** [comment]
Logo<br />**Checkmarx SAST - Scan Summary & Details**
### Cx-SAST Summary
Total of 4 vulnerabilities<br />High&nbsp;**0 High**<br />Medium&nbsp;**3 Medium**<br />Low&nbsp;**1 Low**<br />Info&nbsp;**0 Info**<br />
### Violation Summary
Medium&nbsp;**3** **Medium**<br />
[View more details on Checkmarx UI](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1049057&projectid=728)
### Cx-SAST Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---
[30](https://github.com/Checkmarx/kics/blob/feature/filter_and_alarm_exist_for_aws_management_console_authentication_failures/cmd/builder/main.go#L30) |Medium|Path_Traversal|cmd/builder/main.go|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1049057&projectid=728&pathid=3)
[52](https://github.com/Checkmarx/kics/blob/feature/filter_and_alarm_exist_for_aws_management_console_authentication_failures/pkg/kics/service.go#L52) |Medium|Denial_Of_Service_Resource_Exhaustion|pkg/kics/service.go|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1049057&projectid=728&pathid=1)
[80](https://github.com/Checkmarx/kics/blob/feature/filter_and_alarm_exist_for_aws_management_console_authentication_failures/test/main_test.go#L80) |Medium|Denial_Of_Service_Resource_Exhaustion|test/main_test.go|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1049057&projectid=728&pathid=2)

**Review by rogeriopeixotocx — APPROVED** [review]
LGTM

**Pull request merged**